pygmt/helpers/utils.py

```diff
@@ -14,7 +14,63 @@
 from pygmt.exceptions import GMTInvalidInput
 
 
-def data_kind(data, x=None, y=None, z=None, required_z=False):
+def _validate_data_input(data=None, x=None, y=None, z=None, required_z=False):
+    """
+    Check if the combination of data/x/y/z is valid.
+
+    Examples
+    --------
+    >>> _validate_data_input(data="infile")
+    >>> _validate_data_input(x=[1, 2, 3], y=[4, 5, 6])
+    >>> _validate_data_input(x=[1, 2, 3], y=[4, 5, 6], z=[7, 8, 9])
+    >>> _validate_data_input()
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: No input data provided.
+    >>> _validate_data_input(x=[1, 2, 3])
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Must provide both x and y.
+    >>> _validate_data_input(y=[4, 5, 6])
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Must provide both x and y.
+    >>> _validate_data_input(x=[1, 2, 3], y=[4, 5, 6], required_z=True)
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Must provide x, y, and z.
+    >>> _validate_data_input(data="infile", x=[1, 2, 3])
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Too much data. Use either data or x/y/z.
+    >>> _validate_data_input(data="infile", y=[4, 5, 6])
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Too much data. Use either data or x/y/z.
+    >>> _validate_data_input(data="infile", z=[7, 8, 9])
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Too much data. Use either data or x/y/z.
+
+    Raises
+    ------
+    GMTInvalidInput
+        If the data input is not valid.
+    """
+    if data is None:  # data is None
+        if x is None and y is None:  # both x and y are None
+            raise GMTInvalidInput("No input data provided.")
+        elif x is None or y is None:  # either x or y is None
+            raise GMTInvalidInput("Must provide both x and y.")
+        else:  # both x and y are not None
+            if required_z and z is None:
+                raise GMTInvalidInput("Must provide x, y, and z.")
+    else:  # data is not None
+        if x is not None or y is not None or z is not None:
+            raise GMTInvalidInput("Too much data. Use either data or x/y/z.")
```

Comment:
> It makes more sense to also move the following lines to the `_validate_data_input` function, but it's not easy.
> 
> https://github.com/GenericMappingTools/pygmt/blob/cca17c74d2b36e6cee0098e81b0c75d3d2a8f201/pygmt/helpers/utils.py#L83-L87

Hmm, can't we just add it here?

```suggestion
    else:  # data is not None
        if x is not None or y is not None or z is not None:
            raise GMTInvalidInput("Too much data. Use either data or x/y/z.")
        if required_z and (
            getattr(data, "shape", (3, 3))[1] < 3  # np.array, pd.DataFrame
            or len(getattr(data, "data_vars", (0, 1, 2))) < 3  # xr.Dataset
        ):
            raise GMTInvalidInput("data must provide x, y, and z columns.")
```

Reply:
> But maybe we can do `hasattr(data, "shape") or hasattr(data, "data_vars")` first?

Sounds promising.

Reply:
Haven't tested this, but maybe something like this:

```suggestion
    else:  # data is not None
        if x is not None or y is not None or z is not None:
            raise GMTInvalidInput("Too much data. Use either data or x/y/z.")
        if (
            hasattr(data, "shape")  # np.array, pd.DataFrame
            and getattr(data, "shape", (3, 3))[1] < 3
        ) or (
            hasattr(data, "data_vars")  # xr.Dataset
            and len(getattr(data, "data_vars", (0, 1, 2))) < 3
        ):
            raise GMTInvalidInput("data must provide x, y, and z columns.")
```

Not sure why I added a default value with `getattr` at https://github.com/GenericMappingTools/pygmt/pull/1523, might be able to remove it if not needed, but test first.

Reply:
> Can't remember why I added a default value with `getattr` at #1523,

`getattr(data, "shape", (3, 3))` returns `(3,3)` if `data` doesn't has the `shape` attribute, so it won't raise the exception.

With the above changes, `test_geopandas_plot3d_default_cube` will fail, because in this test: 
```
>>> multipoint = shapely.geometry.MultiPoint([(0.5, 0.5, 0.5), (1.5, 1.5, 1.5)])
>>> gdf = gpd.GeoDataFrame(geometry=[multipoint])
>>> gdf.shape
(1, 1)
```
`gdf` has a shape `(1,1)`.

Reply:
Add a `not hasattr(data, "__geo_interface__")` check perhaps?

Reply:
In ca2d49f2dbfd1daafb492f7032d45a04c791bba4, I changed the order of some codes to determine the data kind first and then check if the data input is valid. It works well.

Reply:
Nice, that works too. Can you fix the pylint error? Should be good after that.